README.md
# tidySpectR

![R-CMD-check](https://github.com/CVUA-RRW/tidySpectR/workflows/R-CMD-check/badge.svg)
[![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)


## Introduction

The aim of this package is to provide tidy methods for the 
analysis of spectral dataset in order to improve 
accessibility and reproducibility of chemometric analysis.

This package is meant to fill the gap between data acquisition 
and preprocessing (phasing, baseline correction) and modelling
with the tidymodels metapackage.

It provides a framework for processing and analysis of spectral data:
 * A `collection` class storing spectra informations and providing methods for 
spectral region selection, normalization and bucketting.
 * A `processing_template` class storing processing steps and allowing to repeat
these steps on other `collection` objects.
 * A few `step_` methods implementing some data preprocessing methods for modelling and/or 
multivariate data analysis, to use with the [tidymodels recipe package](https://github.com/tidymodels/recipes). 

## Installation

This package is not yet on CRAN, install from github:

```r
require("devtools")
install_github("CVUA-RRW/tidySpectR")
```

## Quickstart

### Processing spectral data

```r
library(tidyverse)
library(tidySpectR)
library(tidymodels)

data(fa_nmr)

print(fa_nmr)
#> Spectra collection containing 6 entries.
#> Number of bins: 2250
#> Limits: -2.123166e-05 8.999979
#> Labels: conventional organic
#> 
#> Processing:

processed <- fa_nmr %>% 
             # Trimming edges
             extract(from = 0.5, to = 7.2) %>% 
             # Masking water peak
             mask(from = 4.58, to = 4.68) %>% 
             # Masking MeOD peak
             mask(from = 3.33, to = 3.38) %>% 
             # Bucketting with a bucket size of 0.04 ppm
             bucket_uniform(width = 0.04) %>% 
             # Normalizing with the intensity of the methylen group of phospahtidylcholin
             normalize_internalStandard(from = 3.58, to = 3.64) 

print(processed)
#> Spectra collection containing 6 entries.
#> Number of bins: 166
#> Limits: 0.4999788 7.203979
#> Labels: conventional organic
#> 
#> Processing:
#> Step 1 / 5 : extract
#> Step 2 / 5 : mask
#> Step 3 / 5 : mask
#> Step 4 / 5 : uniform_binning
#> Step 5 / 5 : internalStandard_normalization
```

### Reproducing processing on new data

```recipe
# Export processor
processor <- export_processor(processed)

# Apply processor on new data
new_data <- process(processor, fa_nmr)

print(new_data)
#> Spectra collection containing 6 entries.
#> Number of bins: 166
#> Limits: 0.4999788 7.203979
#> Labels: conventional organic
#> 
#> Processing:
#> Step 1 / 5 : extract
#> Step 2 / 5 : mask
#> Step 3 / 5 : mask
#> Step 4 / 5 : cutsom_bucketting
#> Step 5 / 5 : factor_normalization
```

### Streamlining to tidymodels

```r
# Exporting processed spectra as a data matrix
df <- tidy(processed)

print(df[,1:5]) 
#> # A tibble: 6 x 5
#>   id         label      `0.51997876833874… `0.55997876833874… `0.59997876833874…
#>   <chr>      <fct>                   <dbl>              <dbl>              <dbl>
#> 1 201981241… organic                 0.676              0.658               4.63
#> 2 201981241… organic                 0.722              0.711               5.20
#> 3 201981241… organic                 0.588              0.544               3.90
#> 4 201993059… conventio…              0.631              0.657               4.74
#> 5 201993062… conventio…              0.565              0.540               4.11
#> 6 201995046… conventio…              0.689              0.671               5.31

# Pre-processing for data analysis using the `recipe` package
nmr_preprocessing <- recipe(df, label ~.) %>%
                     update_role(id, new_role = "ID") %>% 
                     # Pareto-VAST scaling - Implemented in this package
                     step_vast(all_predictors(), scaling = 'pareto') %>%
                     # PCA transformation (from the recipes package)
                     step_pca(all_predictors(), num_comp = 5)

# Check the data 
nmr_preprocessing %>% prep() %>% juice()
#> # A tibble: 6 x 7
#>   id          label            PC1    PC2    PC3    PC4    PC5
#>   <fct>       <fct>          <dbl>  <dbl>  <dbl>  <dbl>  <dbl>
#> 1 20198124123 organic       -0.234   8.81 10.4   -0.197  4.37
#> 2 20198124124 organic      -39.6     9.30 -9.34  -2.50   0.787
#> 3 20198124125 organic       32.6    -1.86  0.172 -9.63  -2.44
#> 4 20199305928 conventional   5.01    6.60  2.28   6.68  -5.72
#> 5 20199306281 conventional  37.7    -5.63 -6.97   5.22   3.38
#> 6 20199504645 conventional -35.4   -17.2   3.43   0.440 -0.371
```

## Contributions

Contributions are very welcome!
If you would like to see new functionnalities submit an issue directly 
on github or start a PR.

NEWS.md
# tidySpectR (development version)

## Version 0.2.1
 * Added the `processing_step` and `processing_template` classes and associated methods to store 
function calls to processing steps of `collection` objects.
 * Added the $processor argument to `collection` objects to store a `processing_template`.
 * Added methods to export and reseuse processors from `collection` objects

## Version 0.1.4

 * Fixed an issue caused by large slackness in `bucket_optimized`
 * Added probabilistic quotient normalization method
 * `add_labels` now only add labels to existing IDs
 * Added a SNR argument to `bucket_aibin` to eliminate over-bucketting of noise in the spectral regions
 * Supressed printing of QC results inherited from the `opls`function in the `step_opls_denoise` method
 * Added `median_spectrum` method

## Version 0.1.3

 * Ported bucket_aibin code to C++ for performance increase

## Version 0.1.2

 * Added a `subsample´ method to extract a subset of spectra from a collection
 * Added the `pull_limits`getter to retrieve X-axis limits
 * Added `step_opls_denoise` to filter structural noise

## Version 0.1.1

 * Added `add_labels` to reference
 * Fixed `normalize_aibin`
 * All `bucket_` methods now support multiprocessing. 
 * In order to better integrate multiprocessing, Cluster definition has been removed from ther methods body. Cluster now need to be defined manually be the user (see for example `bucket_aibin` documentation.

## Version 0.1.0

First release!

R/utils.R
# Misc functions for the handling of collection objects

#' Reduce the resolution of Spectra
#'
#' @export
#' @aliases reduce_resolution reduce_resolution.collection
#' @param x A`collection` object to normalize
#' @param ... further arguments passed to or from other methods(not
#'   currently used).
reduce_resolution <- function(x, ...)
    UseMethod("reduce_resolution")

#' @rdname reduce_resolution
#' @param k an integer expressing the reduction in resolution.
#'   For example `k=2`will keep every second point, `k=3` every third.
#' @note Reducing resolution of the spectra before computation intensive steps 
#'   like AI binning can significantly improve their speed, while minimally affecting the output.
#'
#' You should however be careful with the choice of k and in the caase of subsequent binning 
#'   make sure to remove the outter limits of the returned results as these will not match
#'   the outter limits of the full resolution spectra.
#' 
#' @examples
#' library(tidySpectR)
#' # Speeding up AI binning
#' reduced <- reduce_resolution(fa_nmr, k=2)
#' binned <- bucket_aibin(extract(reduced,0.5, 7.2), 
#'                        0.5, 
#'                        mask(reduced, -Inf, 8))
#' breaks <- pull_breaks(binned)
#' 
#' # last break does not match spectra limit, leaving it so would create 
#' # a bin at the ends of the spectra
#' breaks[1]
#' breaks[length(breaks)]
#' pull_limits(extract(fa_nmr, 0.5, 7.2))
#'
#' breaks <- breaks[-1] # remove first break
#' breaks <- breaks[-length(breaks)] # remove last break
#' processed <- bucket_from_breaks(extract(fa_nmr, 0.5, 7.2), breaks)
#'
#' # Full resolution bucketted spectra
#' processed
#' @export
#' @importFrom dplyr slice row_number
reduce_resolution.collection <- function(x, k, ...){
    new_obj <- x
    new_obj$data <- data2wide(x$data) %>%
           slice(which(row_number() %% k == 1)) %>%
           wide2long()
    return(new_obj)
}

#' Calculate the average spectrum 
#'
#' Calculates median of spectra 
#' 
#' @aliases median_spectrum median_spectrum.collection
#' @export
median_spectrum <- function(x, ...)
    UseMethod("median_spectrum")

#' @rdname median_spectrum
#' @param x A`collection` object
#' @param group How to perform grouping. Either `all`or `labels`
#' @param ... further arguments passed to or from other methods(not
#'   currenctly used).
#' @return An updated version of `collection`.
#'
#' @importFrom dplyr group_by summarise mutate select arrange inner_join
#' @importFrom tibble add_column tibble
#' @importFrom stats median
#' @export
#' @examples
#' library(tidySpectR)
#'
#' median_spectrum(fa_nmr)
#'
#' median_spectrum(fa_nmr, group = 'labels')
median_spectrum.collection <- function(x, group = 'all', ...){
    new_obj <- x
    if (group == 'all'){
        new_obj$data <- x$dat %>% 
                        group_by(bins, bin_start, bin_end) %>%
                        summarise(values = median(values), .groups= 'drop') %>%
                        add_column(id = "all", .before = 1) %>% 
                        mutate(id = as.factor(id))
                        
        new_obj$labels <- tibble(id = "all", label = NA) %>%
                          mutate(id = as.factor(id),
                                 label = as.factor(label))
                                 
    } else if (group == 'labels'){
        new_obj$data <- x$dat %>% 
                        inner_join(x$labels, by = 'id') %>%
                        group_by(bins, bin_start, bin_end, label) %>%
                        summarise(values = median(values), .groups= 'drop') %>%
                        mutate(id = label, .before = 1) %>%
                        select(-label) %>%
                        arrange(id, bins)
        
        new_obj$labels <- tibble(id = unique(new_obj$data$id),
                                 label = NA) %>%
                          mutate(id = as.factor(id),
                                 label = as.factor(label))
    
    } else{
        rlang::abort("Invalid value for `group`, use either `all` or `labels`")
    }
    
    return(new_obj)
}

#' Calculate the average spectrum 
#'
#' Averages spectra 
#' 
#' @aliases average_spectrum average_spectrum.collection
#' @export
average_spectrum <- function(x, ...)
    UseMethod("average_spectrum")

#' @rdname average_spectrum
#' @param x A`collection` object
#' @param group How to perform grouping. Either `all`or `labels`
#' @param ... further arguments passed to or from other methods(not
#'   currenctly used).
#' @return An updated version of `collection`.
#'
#' @importFrom dplyr group_by summarise mutate select arrange inner_join
#' @importFrom tibble add_column tibble
#' @export
#' @examples
#' library(tidySpectR)
#'
#' average_spectrum(fa_nmr)
#'
#' average_spectrum(fa_nmr, group = 'labels')
average_spectrum.collection <- function(x, group = 'all', ...){
    new_obj <- x
    if (group == 'all'){
        new_obj$data <- x$dat %>% 
                        group_by(bins, bin_start, bin_end) %>%
                        summarise(values = mean(values), .groups= 'drop') %>%
                        add_column(id = "all", .before = 1) %>% 
                        mutate(id = as.factor(id))
                        
        new_obj$labels <- tibble(id = "all", label = NA) %>%
                          mutate(id = as.factor(id),
                                 label = as.factor(label))
                                 
    } else if (group == 'labels'){
        new_obj$data <- x$dat %>% 
                        inner_join(x$labels, by = 'id') %>%
                        group_by(bins, bin_start, bin_end, label) %>%
                        summarise(values = mean(values), .groups= 'drop') %>%
                        mutate(id = label, .before = 1) %>%
                        select(-label) %>%
                        arrange(id, bins)
        
        new_obj$labels <- tibble(id = unique(new_obj$data$id),
                                 label = NA) %>%
                          mutate(id = as.factor(id),
                                 label = as.factor(label))
    
    } else{
        rlang::abort("Invalid value for `group`, use either `all` or `labels`")
    }
    
    return(new_obj)
}

#' Pull the bin (or bucket) number of a spectra collection
#'
#' For a collection, retrieve the number of bins.
#' 
#' @aliases pull_numbin pull_numbin.collection
#' @export
pull_numbin <- function(x, ...)
    UseMethod("pull_numbin")

#' @rdname pull_numbin
#' @param x A spectra collection
#' @param ... further arguments passed to or from other methods(not
#'   currenctly used).
#' @return A tibble.
#' @importFrom dplyr group_by summarise n
#' @export
#' @examples
#' library(tidySpectR)
#' pull_numbin(fa_nmr)
pull_numbin.collection <- function(x, ...){
    x$data %>% 
        group_by(id) %>%
        summarise("numbin"= n(), .groups = "drop")
}

#' Pull the breaking points (or bin limits) of a spectra collection
#'
#' For a collection, retrieve the bin limits, including the edges 
#'   (min and max limits of the spectra).
#'
#' @aliases pull_breaks pull_breaks.collection
#' @export
pull_breaks <- function(x, ...)
    UseMethod("pull_breaks")

#' @rdname pull_breaks
#' @param x A spectra collection
#' @param ... further arguments passed to or from other methods(not
#'   currenctly used).
#' @return A vector of dbl.
#' @importFrom dplyr pull
#' @export
#' @examples
#' library(tidySpectR)
#' pull_breaks(fa_nmr) %>% head()
pull_breaks.collection <- function(x, ...){
    breaks_low <- x$data %>%
                pull(bin_start) %>% 
                sort() %>%
                unique()
    highest <- x$data %>%
                pull(bin_end) %>%
                unique() %>%
                sort() %>%
                last()
    breaks_low %>% 
        append(highest) %>%
        sort()
}

#' Pull the ids of a spectra collection
#'
#' For a collection, retrieve the sample identifiers 
#' 
#' @aliases pull_ids pull_ids.collection
#' @export
pull_ids <- function(x, ...)
    UseMethod("pull_ids")

#' @rdname pull_ids
#' @param x A spectra collection
#' @param ... further arguments passed to or from other methods(not
#'   currenctly used).
#' @return A vector of chr.
#' @export
#' @examples
#' library(tidySpectR)
#' pull_ids(fa_nmr)
pull_ids.collection <- function(x, ...){
    x$data$id %>% unique() %>% as.character()
}

#' Pull the limits of a spectra collection
#'
#' For a collection, the upper and lower limits of the x-axis
#' 
#' @aliases pull_limits pull_limits.collection
#' @export
pull_limits <- function(x, ...)
    UseMethod("pull_limits")

#' @rdname pull_limits
#' @param x A spectra collection
#' @param ... further arguments passed to or from other methods(not
#'   currenctly used).
#' @return A vector of int.
#' @export
#' @examples
#' library(tidySpectR)
#' pull_limits(fa_nmr)
pull_limits.collection <- function(x, ...){
    breaks <- pull_breaks(x) 
    return(c(first(breaks), last(breaks)))
}
    
#' Adds labels 
#' 
#' Add labels to an existing collection by providing a 
#'   dat frame or tibble associating sample ids to labels
#'
#' @aliases add_labels add_labels.collection
#' @export
add_labels <- function(x, ...)
    UseMethod("add_labels")

#' @rdname add_labels
#' @param x A `collection` object 
#' @param labels A dataframe or tibble associating sample ids to labels
#' @param ids_from Name of the id column
#' @param labels_from Name of the label column
#' @param ... further arguments passed to or from other methods(not
#'   currenctly used).
#' @return An updated version of `collection`.
#' @details Pre-existing labels will be overwritten.
#' @importFrom tibble tibble
#' @importFrom dplyr pull mutate semi_join
#' @export
#' @examples
#' library(tidySpectR)
#' library(tibble)
#'
#' newlabs <- tibble(names = pull_ids(fa_nmr),
#'                   conditions = c("control", "control", "control", 
#'                                  "treated", "treated", "treated"))
#'
#' add_labels(fa_nmr, newlabs, ids_from= "names", labels_from= "conditions")
add_labels.collection <- function(x, labels, ids_from, labels_from, ...){
    new_obj <- x
    ids <- x$data %>% 
           pull(id) %>%
           unique() %>%
           tibble(id = .)
           
    new_obj$labels <- tibble(id = pull(labels, var = ids_from),
                             label = pull(labels, var = labels_from)) %>%
                      mutate(id = as.factor(id),
                             label = as.factor(label)) %>%
                      semi_join(ids, by = "id")
    
    return(new_obj)
}
